hollow/src/main/java/com/netflix/hollow/core/index/FieldPath.java

```diff
@@ -203,38 +312,47 @@ protected String getLastRefTypeInPath() {
         } else {
             // Object type
             HollowObjectSchema objectSchema = (HollowObjectSchema) schema;
-            HollowObjectTypeReadState objectTypeReadState = (HollowObjectTypeReadState) typeReadState;
-
-            if (fieldTypes[fieldIndex].equals(HollowObjectSchema.FieldType.REFERENCE)) {
-                int refOrdinal = objectTypeReadState.readOrdinal(ordinal, fieldPositions[fieldIndex]);
-                String refType = objectSchema.getReferencedType(fieldPositions[fieldIndex]);
-                return findValuesFollowingPath(refOrdinal, refType, fieldIndex + 1);
-            } else {
-                switch (fieldTypes[fieldIndex]) {
-                    case INT:
-                        values = new Integer[]{objectTypeReadState.readInt(ordinal, fieldPositions[fieldIndex])};
-                        break;
-                    case LONG:
-                        values = new Long[]{objectTypeReadState.readLong(ordinal, fieldPositions[fieldIndex])};
-                        break;
-                    case DOUBLE:
-                        values = new Double[]{objectTypeReadState.readDouble(ordinal, fieldPositions[fieldIndex])};
-                        break;
-                    case FLOAT:
-                        values = new Float[]{objectTypeReadState.readFloat(ordinal, fieldPositions[fieldIndex])};
-                        break;
-                    case BOOLEAN:
-                        values = new Boolean[]{objectTypeReadState.readBoolean(ordinal, fieldPositions[fieldIndex])};
-                        break;
-                    case STRING:
-                        values = new String[]{objectTypeReadState.readString(ordinal, fieldPositions[fieldIndex])};
-                        break;
-                    default:
-                        throw new IllegalStateException("Invalid field type :" + fieldTypes[fieldIndex] + " cannot read values for this type");
+            HollowObjectTypeDataAccess objectTypeDataAccess = (HollowObjectTypeDataAccess) typeDataAccess;
 
+            if (fieldTypes[fieldIndex].equals(FieldType.REFERENCE)) {
+                int refOrdinal = objectTypeDataAccess.readOrdinal(ordinal, fieldPositions[fieldIndex]);
+                if (refOrdinal >= 0) {
+                    String refType = objectSchema.getReferencedType(fieldPositions[fieldIndex]);
+                    return getAllValues(refOrdinal, refType, fieldIndex + 1);
                 }
+                return new Object[]{};
+            } else {
+                return new Object[]{readFromObject(objectTypeDataAccess, ordinal, fieldTypes[fieldIndex], fieldPositions[fieldIndex])};
             }
         }
         return values;
     }
+
+    private Object readFromObject(HollowObjectTypeDataAccess objectTypeDataAccess, int ordinal, FieldType fieldType, int fieldPosition) {
```

Comment:
This seems like common code and it does not seem to handle checking for special values.

See HollowReadFieldUtils.fieldValueObject to determine whether it makes sense to reuse that.